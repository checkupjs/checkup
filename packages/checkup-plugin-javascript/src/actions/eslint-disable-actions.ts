import { TaskResult, ActionsEvaluator } from '@checkup/core';

export function evaluateActions(taskResult: TaskResult) {
  let actionsEvaluator = new ActionsEvaluator();
  let summaryResult = taskResult.data[0];
  let eslintDisableUsages = summaryResult.count;

  actionsEvaluator.add({
    name: 'reduce-eslint-disable-usages',
    summary: 'Reduce number of eslint-disable usages',
    details: `${eslintDisableUsages} usages of template-lint-disable`,
    defaultThreshold: 2,
    items: [`Total eslint-disable usages: ${eslintDisableUsages}`],
    input: eslintDisableUsages,
  });

  return actionsEvaluator.evaluate(taskResult.config);
}
