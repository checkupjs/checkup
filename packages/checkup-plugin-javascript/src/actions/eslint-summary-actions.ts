import { TaskResult, ActionsEvaluator, MultiValueResult } from '@checkup/core';

export function evaluateActions(taskResult: TaskResult) {
  let actionsEvaluator = new ActionsEvaluator();

  let errors = taskResult.data.find((result: MultiValueResult) => result.key === 'eslint-errors')!;
  let warnings = taskResult.data.find(
    (result: MultiValueResult) => result.key === 'eslint-warnings'
  )!;

  let errorCount = errors.dataSummary.total;
  let warningCount = warnings.dataSummary.total;

  actionsEvaluator.add({
    name: 'reduce-eslint-errors',
    summary: 'Reduce number of eslint errors',
    details: `${errorCount} total errors`,
    defaultThreshold: 20,
    items: [`Total eslint errors: ${errorCount}`],
    input: errorCount,
  });

  actionsEvaluator.add({
    name: 'reduce-eslint-warnings',
    summary: 'Reduce number of eslint warnings',
    details: `${warningCount} total warnings`,
    defaultThreshold: 20,
    items: [`Total eslint warnings: ${warningCount}`],
    input: warningCount,
  });

  return actionsEvaluator.evaluate(taskResult.config);
}
