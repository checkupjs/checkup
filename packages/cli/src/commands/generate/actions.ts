import * as yargs from 'yargs';
import Generator from '../../api/generator';
import { consoleWriter } from '../../checkup';

export const generateActionsCommand = {
  command: 'actions <name> [options]',
  describe: 'Generates checkup actions within a project',
  builder: (yargs: any) => {
    return yargs
      .positional('name', {
        description: 'Name of the actions (foo-task-actions)',
        default: '',
      })
      .options({
        defaults: {
          alias: 'd',
          description: 'Use defaults for every setting',
          boolean: true,
        },
        path: {
          alias: 'p',
          default: '.',
          description: 'The path referring to the directory that the generator will run in',
        },
      });
  },
  handler: async (argv: yargs.Arguments) => {
    try {
      let generator = new Generator({
        path: argv.path as string,
        generator: 'actions',
        name: argv.name as string,
        defaults: argv.defaults as boolean,
      });

      await generator.run();
    } catch (error) {
      consoleWriter.error(error);
    }
  },
};
