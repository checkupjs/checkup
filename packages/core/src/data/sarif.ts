import { Result, Location, Notification } from 'sarif';
import { LintResult, Task, TaskListError } from '../types/tasks';

export const NO_RESULTS_FOUND = 'No results found';

type SarifTaskIdentifier = Pick<Task, 'taskName' | 'taskDisplayName' | 'category' | 'group'>;

/**
 * Builds SARIF Results from a list of LintResults.
 *
 * @param taskContext {Task} This is used to set Task properties on the Result
 * @param lintResults {LintResult[]} The LintResults used to create Result
 * @param [additionalData] {Object} Any additional data to be put into the properties bag
 * @param [customMessages] {Record<string, string>} Custom messages to be rendered for each lintRule
 */
function fromLintResults(
  taskContext: SarifTaskIdentifier,
  lintResults: LintResult[],
  additionalData: object = {},
  customMessages: Record<string, string> = {}
): Result[] {
  if (lintResults.length === 0) {
    return buildEmptyResult(taskContext);
  }

  // When files disable lint rules that are not defined in the context of our Task's analyzer, it adds a lintResult
  // that indicates that the rule was not found. Since this is not an actual error || warning, we filter them out
  const lintRuleNotDefinedRegex = new RegExp('Definition for rule .* was not found');

  return lintResults
    .filter((lintResult) => {
      return lintRuleNotDefinedRegex.test(lintResult.message) === false;
    })
    .map((lintResult) => {
      let message =
        (lintResult.lintRuleId && customMessages[lintResult.lintRuleId]) ?? lintResult.message;

      return {
        locations: buildLocationDataFromLintResult(lintResult),
        message: { text: message },
        occurrenceCount: 1,
        ruleId: taskContext.taskName,
        properties: {
          ...{
            taskDisplayName: taskContext.taskDisplayName,
            category: taskContext.category,
            group: taskContext.group,
          },
          ...additionalData,
          lintRuleId: lintResult.lintRuleId,
        },
      };
    });
}

/**
 * Builds SARIF Results from a list of locations.
 *
 * @param taskContext {Task} This is used to set Task properties on the Result
 * @param locations {string[]} The paths used to create Result
 * @param message {string} The message that identifies the data represented in the Result
 */
function fromLocations(
  taskContext: SarifTaskIdentifier,
  locations: string[],
  message: string
): Result[] {
  if (locations.length === 0) {
    return buildEmptyResult(taskContext, message);
  }

  return locations.map((path) => {
    return {
      locations: buildLocationDataFromPath(path),
      message: { text: message },
      occurrenceCount: 1,
      ruleId: taskContext.taskName,
      properties: {
        taskDisplayName: taskContext.taskDisplayName,
        category: taskContext.category,
        group: taskContext.group,
      },
    };
  });
}

export function fromTaskErrors(errors: TaskListError[]): Notification[] {
  return errors.map((error) => {
    return {
      message: { text: error.error.message },
      associatedRule: {
        id: error.taskName,
      },
      properties: { fullError: error.error.stack },
    };
  });
}

export const sarifBuilder = {
  fromLintResults,
  fromLocations,
  notifications: {
    fromTaskErrors,
  },
};

function buildEmptyResult(taskContext: SarifTaskIdentifier, consoleMessage?: string): Result[] {
  return [
    {
      message: { text: NO_RESULTS_FOUND },
      ruleId: taskContext.taskName,
      properties: {
        consoleMessage,
        ...{
          taskDisplayName: taskContext.taskDisplayName,
          category: taskContext.category,
          group: taskContext.group,
        },
      },
    },
  ];
}

/**
 * @param lintResult {LintResult}
 * @returns Location[]
 */
function buildLocationDataFromLintResult(lintResult: LintResult): Location[] {
  const location: Location = {
    physicalLocation: {
      artifactLocation: {
        uri: lintResult.filePath,
      },
    },
  };

  if (
    (lintResult.line > 0 || lintResult.column > 0) &&
    location &&
    location.physicalLocation !== undefined
  ) {
    location.physicalLocation.region = {};
    if (lintResult.line > 0) {
      location.physicalLocation.region.startLine = lintResult.line;
    }
    if (lintResult.column > 0) {
      location.physicalLocation.region.startColumn = lintResult.column;
    }
  }

  return [location];
}

/**
 * @param paths {string[]}
 * @returns Location[]
 */
function buildLocationDataFromPath(path: string): Location[] {
  return [
    {
      physicalLocation: {
        artifactLocation: {
          uri: path,
        },
      },
    },
  ];
}
