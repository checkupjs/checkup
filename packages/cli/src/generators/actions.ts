import * as _ from 'lodash';
import * as chalk from 'chalk';
import * as t from '@babel/types';
import * as recast from 'recast';

import traverse from '@babel/traverse';
import { Answers } from 'inquirer';
import BaseGenerator, { Works } from './base-generator';
import { Options } from '../commands/generate';
import { join } from 'path';
import { AstTransformer, CheckupError } from '@checkup/core';

interface ActionOptions extends Options {
  taskName: string;
  pascalCaseName: string;
  typescript: boolean;
}

export default class ActionsGenerator extends BaseGenerator {
  works: Works = Works.InsidePlugin;
  answers!: Answers;

  constructor(args: any, public options: ActionOptions) {
    super(args, options);
  }

  initializing() {
    if (!this.canRunGenerator) {
      throw new CheckupError(
        `Can only generate actions from inside a Checkup plugin directory`,
        `Run ${chalk.bold.white(
          'checkup generate actions'
        )} from the root of a Checkup plugin or use the ${chalk.bold.white(
          '--path'
        )} option to specify the path to a Checkup plugin`
      );
    }
  }

  async prompting() {
    this.headline(`${this.options.name}-actions`);

    const defaults = {
      typescript: true,
    };

    if (this.options.defaults) {
      this.answers = defaults;
    } else {
      this.answers = await this.prompt([
        {
          type: 'confirm',
          name: 'typescript',
          message: 'TypeScript',
          default: () => true,
        },
        {
          type: 'input',
          name: 'taskName',
          message: `Enter the task name that these actions will be associated with (the string value in the taskName property of the task class).`,
        },
      ]);
    }

    this.options.taskName = this.answers.taskName;
    this.options.pascalCaseName = _.upperFirst(_.camelCase(this.options.name));
    this.options.typescript = this.answers.typescript;
  }

  writing() {
    this.sourceRoot(join(__dirname, '../../templates/src/actions'));

    if (
      !this.fs.exists(
        this.destinationPath(`${this._dir}/registrations/register-actions.${this._ext}`)
      )
    ) {
      this.fs.copy(
        this.templatePath(`src/registrations/register-actions.${this._ext}.ejs`),
        this.destinationPath(`${this._dir}/registrations/register-actions.${this._ext}`)
      );
    }

    this.fs.copyTpl(
      this.templatePath(`src/actions/actions.${this._ext}.ejs`),
      this.destinationPath(`${this._dir}/actions/${this.options.name}-actions.${this._ext}`),
      this.options
    );

    this._transformHooks();
  }

  private _transformHooks() {
    let hooksDestinationPath = this.destinationPath(
      `${this._dir}/registrations/register-actions.${this._ext}`
    );

    let registerActionsSource = this.fs.read(hooksDestinationPath);
    let registerActionStatement = t.expressionStatement(
      t.callExpression(t.identifier('registerActions'), [
        t.stringLiteral(this.options.taskName),
        t.identifier(`evaluate${this.options.pascalCaseName}Actions`),
      ])
    );

    let newActionImportSpecifier = t.importSpecifier(
      t.identifier(`evaluate${this.options.pascalCaseName}Actions`),
      t.identifier('evaluateActions')
    );
    let tasksImportDeclaration: t.ImportDeclaration = t.importDeclaration(
      [newActionImportSpecifier],
      t.stringLiteral(`../actions/${this.options.name}-actions`)
    );

    let code = new AstTransformer(registerActionsSource, recast.parse, traverse, {
      parser: require('recast/parsers/typescript'),
    })
      .traverse({
        Program(path) {
          path.node.body.splice(1, 0, tasksImportDeclaration);
        },
        BlockStatement(path) {
          path.node.body.push(registerActionStatement);
        },
      })
      .generate();

    this.fs.write(hooksDestinationPath, code);
  }
}
