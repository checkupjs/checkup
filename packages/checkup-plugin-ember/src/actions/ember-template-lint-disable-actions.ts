import { ActionsEvaluator, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let actionsEvaluator = new ActionsEvaluator();
  let templateLintDisableUsages = sumOccurrences(taskResults);

  actionsEvaluator.add({
    taskName: 'ember-template-lint-disable',
    name: 'reduce-template-lint-disable-usages',
    summary: 'Reduce number of template-lint-disable usages',
    details: `${templateLintDisableUsages} usages of template-lint-disable`,
    defaultThreshold: 2,
    items: [`Total template-lint-disable usages: ${templateLintDisableUsages}`],
    input: templateLintDisableUsages,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
