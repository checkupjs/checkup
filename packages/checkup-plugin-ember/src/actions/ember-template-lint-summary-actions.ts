import { ActionsEvaluator, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let actionsEvaluator = new ActionsEvaluator();

  let errors = taskResults.filter((result: Result) => result.properties?.type === 'error')!;
  let warnings = taskResults.filter((result: Result) => result.properties?.type === 'warning')!;

  let errorCount = sumOccurrences(errors);
  let warningCount = sumOccurrences(warnings);

  actionsEvaluator.add({
    taskName: 'ember-template-lint-summary',
    name: 'reduce-template-lint-errors',
    summary: 'Reduce number of template-lint errors',
    details: `${errorCount} total errors`,
    defaultThreshold: 20,
    items: [`Total template-lint errors: ${errorCount}`],
    input: errorCount,
  });

  actionsEvaluator.add({
    taskName: 'ember-template-lint-summary',
    name: 'reduce-template-lint-warnings',
    summary: 'Reduce number of template-lin warnings',
    details: `${warningCount} total warnings`,
    defaultThreshold: 20,
    items: [`Total template-lint warnings: ${warningCount}`],
    input: warningCount,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
