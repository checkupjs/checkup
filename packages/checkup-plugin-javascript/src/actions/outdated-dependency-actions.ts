import { ActionsEvaluator, toPercent, TaskConfig, sumOccurrences } from '@checkup/core';
import { Result } from 'sarif';

export function evaluateActions(taskResults: Result[], taskConfig: TaskConfig) {
  let totalDependencies = sumOccurrences(taskResults);

  let outdatedDependencies = {
    major: sumOccurrences(taskResults.filter((taskResult) => taskResult.message.text === 'major')),
    minor: sumOccurrences(taskResults.filter((taskResult) => taskResult.message.text === 'minor')),
  };

  let actionsEvaluator = new ActionsEvaluator();

  actionsEvaluator.add({
    name: 'reduce-outdated-major-dependencies',
    summary: 'Update outdated major versions',
    details: `${outdatedDependencies.major} major versions outdated`,
    defaultThreshold: 0.05,
    items: [],
    input: outdatedDependencies.major / totalDependencies,
  });
  actionsEvaluator.add({
    name: 'reduce-outdated-minor-dependencies',
    summary: 'Update outdated minor versions',
    details: `${outdatedDependencies.minor} minor versions outdated`,
    defaultThreshold: 0.05,
    items: [],
    input: outdatedDependencies.minor / totalDependencies,
  });
  actionsEvaluator.add({
    name: 'reduce-outdated-dependencies',
    summary: 'Update outdated versions',
    details: `${toPercent(
      (outdatedDependencies.major + outdatedDependencies.minor) / totalDependencies
    )} of versions outdated`,
    defaultThreshold: 0.2,
    items: [],
    input: (outdatedDependencies.major + outdatedDependencies.minor) / totalDependencies,
  });

  return actionsEvaluator.evaluate(taskConfig);
}
