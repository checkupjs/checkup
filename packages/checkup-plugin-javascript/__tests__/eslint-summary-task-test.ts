import { CheckupProject, getTaskContext } from '@checkup/test-helpers';
import {
  EslintSummaryTask,
  readEslintConfig,
  ACCEPTED_ESLINT_CONFIG_FILES,
} from '../src/tasks/eslint-summary-task';
import { evaluateActions } from '../src/actions/eslint-summary-actions';
import { PackageJson } from 'type-fest';
import { getPluginName, Task, FilePathArray } from '@checkup/core';

import { Result, Location } from 'sarif';

describe('eslint-summary-task', () => {
  let project: CheckupProject;
  let pluginName = getPluginName(__dirname);
  let task: Task;
  let result: Result[];

  beforeAll(async () => {
    project = new CheckupProject('checkup-app', '0.0.0');
    project.files['included-path.js'] = `function foo() {
        var whatever = 'ESLint'
        return {
          name: whatever
        };
      }`;
    project.files['excluded-path.js'] = `function shmoo() {
        var whatever = 'ESLint'
        return {
          name: whatever
        };
      }`;
    project.files['.eslintrc.js'] = `module.exports = {
        rules: {
            semi: 'error',
            'no-var': 'warn'
        }
      };`;
    project.writeSync();
    project.gitInit();
    project.install();

    task = new EslintSummaryTask(
      pluginName,
      getTaskContext({
        options: { cwd: project.baseDir },
        pkg: project.pkg,
        paths: project.filePaths.filter((path) => path.includes('included-path')) as FilePathArray,
        config: {
          tasks: {
            'javascript/eslint-summary': [
              'on',
              {
                actions: {
                  'reduce-eslint-errors': ['on', { threshold: 0 }],
                  'reduce-eslint-warnings': ['on', { threshold: 0 }],
                },
              },
            ],
          },
        },
      })
    );
    result = await task.run();
  });

  afterAll(() => {
    project.dispose();
  });

  it('summarizes eslint and outputs to JSON', async () => {
    expect(result).toMatchSnapshot();
  });

  it('only lints the files passed in via paths array', async () => {
    let excludedPathsResults = result.filter(
      (resultData: Result) =>
        (
          resultData.locations?.filter((fileLocation: Location) => {
            return fileLocation?.physicalLocation?.artifactLocation?.uri?.includes('excluded-path');
          }) || []
        ).length > 0
    );

    let includedPathsResults = result.filter(
      (resultData: Result) =>
        (
          resultData.locations?.filter((fileLocation: Location) => {
            return fileLocation?.physicalLocation?.artifactLocation?.uri?.includes('included-path');
          }) || []
        ).length > 0
    );

    expect(excludedPathsResults).toHaveLength(0);
    expect(includedPathsResults).toHaveLength(2);
  });

  it('returns correct action items if there are too many warnings or errors', async () => {
    let actions = evaluateActions(result, task.config);

    expect(actions).toHaveLength(2);
    expect(actions).toMatchInlineSnapshot(`
      Array [
        Object {
          "defaultThreshold": 20,
          "details": "1 total errors",
          "input": 1,
          "items": Array [
            "Total eslint errors: 1",
          ],
          "name": "reduce-eslint-errors",
          "summary": "Reduce number of eslint errors",
        },
        Object {
          "defaultThreshold": 20,
          "details": "1 total warnings",
          "input": 1,
          "items": Array [
            "Total eslint warnings: 1",
          ],
          "name": "reduce-eslint-warnings",
          "summary": "Reduce number of eslint warnings",
        },
      ]
    `);
  });
});

describe('readEslintConfig', () => {
  const eslintConfigJson = 'airbnb';
  const pkg: PackageJson & { eslintConfig: string } = {
    name: 'foo-project',
    version: '0.0.0',
    keywords: [],
    eslintConfig: eslintConfigJson,
  };

  ACCEPTED_ESLINT_CONFIG_FILES.forEach((file) => {
    it(`it returns ${file}, even if there is also a JSON config (as .eslintrc* are prioritized over config in package.json)`, () => {
      let expectedConfig = 'foo';

      let project = new CheckupProject('foo-project');
      project.files[file] = expectedConfig;
      project.writeSync();

      const eslintConfig = readEslintConfig(project.filePaths, project.baseDir, pkg);

      expect(eslintConfig).toEqual(expectedConfig);
      project.dispose();
    });
  });

  it('reads config from package.json, only if there is no .eslintrc*', () => {
    const eslintConfig = readEslintConfig([], '.', pkg);

    expect(eslintConfig).toEqual(eslintConfigJson);
  });

  it('reads the higher prioritzed .eslintrc*', () => {
    let expectedConfig = 'foo';

    let project = new CheckupProject('foo-project');
    project.files['.eslintrc.js'] = expectedConfig;
    project.files['.eslintrc.json'] = 'blue';
    project.writeSync();

    const eslintConfig = readEslintConfig(project.filePaths, project.baseDir, pkg);

    expect(eslintConfig).toEqual(expectedConfig);
    project.dispose();
  });
});
