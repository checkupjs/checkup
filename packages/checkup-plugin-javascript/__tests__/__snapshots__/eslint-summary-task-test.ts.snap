// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eslint-summary-task it summarizes eslint and outputs to JSON 1`] = `
Object {
  "errorCount": 1,
  "fixableErrorCount": 1,
  "fixableWarningCount": 1,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": "",
      "fixableErrorCount": 1,
      "fixableWarningCount": 1,
      "messages": Array [
        Object {
          "column": 9,
          "endColumn": 32,
          "endLine": 2,
          "fix": Object {
            "range": Array [
              25,
              28,
            ],
            "text": "let",
          },
          "line": 2,
          "message": "Unexpected var, use let or const instead.",
          "nodeType": "VariableDeclaration",
          "ruleId": "no-var",
          "severity": 1,
        },
        Object {
          "column": 32,
          "endColumn": 1,
          "endLine": 3,
          "fix": Object {
            "range": Array [
              48,
              48,
            ],
            "text": ";",
          },
          "line": 2,
          "message": "Missing semicolon.",
          "nodeType": "VariableDeclaration",
          "ruleId": "semi",
          "severity": 2,
        },
      ],
      "source": "function foo() {
        var whatever = 'ESLint'
        return {
          name: whatever
        };
      }",
      "warningCount": 1,
    },
    Object {
      "errorCount": 0,
      "filePath": "",
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 1,
}
`;
